File: backend/infrastructure/workshop_scanner.py
"""
Scanner for Steam Workshop mods and local packs.
"""
import os
import re
from pathlib import Path
from typing import List, Optional
from backend.domain.models import Mod

class WorkshopScanner:
    def __init__(self):
        pass

    def scan_workshop(self, workshop_dir: str) -> List[Mod]:
        """
        Discovers all valid .pack mods located in the Steam Workshop content directory.
        """
        mods: List[Mod] = []
        if not workshop_dir or not os.path.exists(workshop_dir):
            return mods

        workshop_path = Path(workshop_dir)
        if not workshop_path.is_dir():
            return mods

        try:
            entries = os.listdir(workshop_path)
        except OSError:
            return mods

        for folder_name in entries:
            folder_path = workshop_path / folder_name
            if not folder_path.is_dir():
                continue

            # Check for publish_data.vdf to find custom Steam title if available
            workshop_title = self._extract_workshop_title(folder_path)

            try:
                files = os.listdir(folder_path)
            except OSError:
                continue

            pack_files = [f for f in files if f.lower().endswith('.pack')]
            for pack in pack_files:
                pack_path = folder_path / pack
                
                # Determine thumbnail
                img_name = self._find_thumbnail_image(files, pack)
                thumb_url = f"/workshop_assets/{folder_name}/{img_name}" if img_name else "/static/gemini-svg.svg"

                # File stats
                file_size = 0
                mtime = 0.0
                try:
                    stats = pack_path.stat()
                    file_size = stats.st_size
                    mtime = stats.st_mtime
                except OSError:
                    pass

                # Derive clean title
                clean_title = workshop_title or pack.removesuffix('.pack').replace('_', ' ').strip()

                mod = Mod(
                    id=folder_name,
                    name=pack,
                    title=clean_title,
                    real_path=str(pack_path.resolve()),
                    thumb=thumb_url,
                    url=f"https://steamcommunity.com/sharedfiles/filedetails/?id={folder_name}",
                    file_size_bytes=file_size,
                    last_modified=mtime
                )
                mods.append(mod)

        # Sort alphabetically by title/name
        mods.sort(key=lambda m: (m.title or m.name).lower())
        return mods

    def _find_thumbnail_image(self, files: List[str], pack_name: str) -> Optional[str]:
        """Finds matching thumbnail image for the pack file."""
        expected_png = pack_name[:-5] + ".png" if pack_name.lower().endswith('.pack') else pack_name + ".png"
        expected_jpg = pack_name[:-5] + ".jpg" if pack_name.lower().endswith('.pack') else pack_name + ".jpg"

        file_set = {f.lower(): f for f in files}

        if expected_png.lower() in file_set:
            return file_set[expected_png.lower()]
        if expected_jpg.lower() in file_set:
            return file_set[expected_jpg.lower()]
        if "thumbnail.png" in file_set:
            return file_set["thumbnail.png"]
        if "thumbnail.jpg" in file_set:
            return file_set["thumbnail.jpg"]
        if "thumb.png" in file_set:
            return file_set["thumb.png"]
        if "thumb.jpg" in file_set:
            return file_set["thumb.jpg"]

        return None

    def _extract_workshop_title(self, folder_path: Path) -> Optional[str]:
        """Attempts to read mod title from publish_data.vdf or metadata if present."""
        vdf_file = folder_path / "publish_data.vdf"
        if vdf_file.exists():
            try:
                content = vdf_file.read_text(encoding='utf-8', errors='ignore')
                match = re.search(r'"title"\s+"([^"]+)"', content, re.IGNORECASE)
                if match:
                    return match.group(1).strip()
            except Exception:
                pass
        return None

File: src-tauri/src/services/workshop_scanner.rs
use std::fs;
use std::path::{Path, PathBuf};
use regex::Regex;
use crate::domain::Mod;

pub struct WorkshopScanner;

impl WorkshopScanner {
    pub fn scan_workshop(workshop_dir: &str) -> Vec<Mod> {
        let mut mods = Vec::new();
        let workshop_path = Path::new(workshop_dir);

        if !workshop_path.is_dir() {
            return mods;
        }

        let entries = match fs::read_dir(workshop_path) {
            Ok(e) => e,
            Err(_) => return mods,
        };

        for entry in entries.flatten() {
            let folder_path = entry.path();
            if !folder_path.is_dir() {
                continue;
            }

            let folder_name = folder_path.file_name().unwrap_or_default().to_string_lossy().to_string();
            let workshop_title = Self::extract_workshop_title(&folder_path);

            let sub_entries = match fs::read_dir(&folder_path) {
                Ok(e) => e,
                Err(_) => continue,
            };

            let mut files: Vec<PathBuf> = Vec::new();
            for sub in sub_entries.flatten() {
                files.push(sub.path());
            }

            let pack_files: Vec<&PathBuf> = files
                .iter()
                .filter(|p| {
                    p.extension()
                        .map_or(false, |ext| ext.to_string_lossy().eq_ignore_ascii_case("pack"))
                })
                .collect();

            for pack_path in pack_files {
                let pack_name = pack_path.file_name().unwrap_or_default().to_string_lossy().to_string();
                let thumb_path = Self::find_thumbnail(&files, &pack_name);
                
                let thumb_url = thumb_path
                    .map(|p| p.to_string_lossy().to_string())
                    .unwrap_or_else(|| "/static/gemini-svg.svg".to_string());

                let mut file_size = 0;
                let mut mtime = 0.0;
                if let Ok(meta) = pack_path.metadata() {
                    file_size = meta.len();
                    if let Ok(modified) = meta.modified() {
                        if let Ok(duration) = modified.duration_since(std::time::UNIX_EPOCH) {
                            mtime = duration.as_secs_f64();
                        }
                    }
                }

                let clean_title = workshop_title.clone().unwrap_or_else(|| {
                    pack_name
                        .trim_end_matches(".pack")
                        .trim_end_matches(".PACK")
                        .replace('_', " ")
                        .trim()
                        .to_string()
                });

                let steam_url = format!("https://steamcommunity.com/sharedfiles/filedetails/?id={}", folder_name);

                mods.push(Mod {
                    id: folder_name.clone(),
                    name: pack_name,
                    title: clean_title,
                    real_path: pack_path.to_string_lossy().to_string(),
                    thumb: thumb_url,
                    url: steam_url,
                    file_size_bytes: file_size,
                    is_movie_pack: false,
                    last_modified: mtime,
                });
            }
        }

        mods.sort_by(|a, b| a.title.to_lowercase().cmp(&b.title.to_lowercase()));
        mods
    }

    fn find_thumbnail(files: &[PathBuf], pack_name: &str) -> Option<PathBuf> {
        let base_stem = pack_name
            .trim_end_matches(".pack")
            .trim_end_matches(".PACK")
            .to_lowercase();

        for file in files {
            if let Some(stem) = file.file_stem() {
                let stem_str = stem.to_string_lossy().to_lowercase();
                if let Some(ext) = file.extension() {
                    let ext_str = ext.to_string_lossy().to_lowercase();
                    if ext_str == "png" || ext_str == "jpg" || ext_str == "jpeg" {
                        if stem_str == base_stem || stem_str == "thumbnail" || stem_str == "thumb" {
                            return Some(file.clone());
                        }
                    }
                }
            }
        }
        None
    }

    fn extract_workshop_title(folder_path: &Path) -> Option<String> {
        let vdf_file = folder_path.join("publish_data.vdf");
        if vdf_file.exists() {
            if let Ok(content) = fs::read_to_string(vdf_file) {
                let re = Regex::new(r#""title"\s+"([^"]+)""#).ok()?;
                if let Some(cap) = re.captures(&content) {
                    if let Some(m) = cap.get(1) {
                        return Some(m.as_str().trim().to_string());
                    }
                }
            }
        }
        None
    }
}
